File: 0337-House Robber III/0337-House Robber III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        vector<int> result = robHelper(root);
        return max(result[0], result[1]);
    }
    
private:
    vector<int> robHelper(TreeNode* root) {
        vector<int> result(2);
        if (root == nullptr)
        {
            return result;
        }
        
        vector<int> left = robHelper(root->left);
        vector<int> right = robHelper(root->right);
        
        result[0] = root->val + left[1] + right[1];
        result[1] = max(left[0], left[1]) + max(right[0], right[1]);
        return result;
    }
};

File: 0301-0400/0337-House Robber III/0337-House Robber III.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def rob(self, root: TreeNode) -> int:
        def dfs(root):
            if root is None:
                return [0, 0]
            
            left = dfs(root.left)
            right = dfs(root.right)
            return root.val + left[1] + right[1], max(left) + max(right)

        return max(dfs(root))
